### PR TITLE
[2.x] Use Laravel's SchemaState for creating dumps

### LINKED ISSUE #42: Use Laravel's *SchemaState for dumping
In order not to have to maintain the routine to create dumps while this functionality already exists in the Laravel framework core, Protector should use the SchemaState classes to create dumps, in this case MySqlSchemaState until Protector supports other DBMS.

I would suggest deriving an own MySqlSchemaState class and overriding the dump method to adjust the parameters accordingly.
This way, other DBMS could be implemented quickly and safely in the future, the same goes for importing the dumps.

### PR DESCRIPTION
**Pull request by lupinitylabs**
This PR resolves #42 by replacing the dump creation logic with using the SchemaState classes from the Laravel framework.
This iteration only supports MySQL databases, and should be used to discuss the approach and get things right before implementing support for PostgreSQL and SQLite.

The default SchemaState classes create dumps that are highly opinionated and do not dump table data, so we have to override the dump method in order to pass our own set of options.

As it turned out in the end, there is not much left that we use of the framework code, but we potentially don't want to pull in any framework changes in the parts we replaced anyways. The approach allows developers to override the classes with their own implementations in case they want to do other fancy stuff when dumping or restoring.

This PR also applies a number of architectural changes, extracting configuration concerns to the HasConfiguration trait to make the Protector class a little more readable. It could be discussed to use a separate configuration class instead, since the trait will not be reused by any other class.

### CODE REVIEW

File: src/Protector.php [hunk 1]
```diff
@@ -520,15 +410,15 @@ public function getMetaData(bool $refresh = false): array
             $this->guardExecEnabled();
 
             $gitInformation = [
-                'gitRevision'     => $this->getGitRevision(),
-                'gitBranch'       => $this->getGitBranch(),
-                'gitRevisionDate' => $this->getGitHeadDate(),
+            'gitRevision'     => $this->getGitRevision(),
+            'gitBranch'       => $this->getGitBranch(),
+            'gitRevisionDate' => $this->getGitHeadDate(),
```
Comment (mszulik) [inline src/Protector.php:415]: indentation

File: src/Protector.php [hunk 2]
```diff
@@ -55,62 +52,20 @@ class Protector
...
 
     /**
      * Imports a specific SQL dump.
      *
-     * @param string $sourceFilePath
-     * @param array $options
+     * @param  string  $sourceFilePath
+     * @param  array  $options
      *
      * @return void
```
Comment (mszulik) [inline src/Protector.php:70]: can be removed too i guess

File: src/Protector.php (hunk 1 above)
Comment (lupinitylabs) [reply src/Protector.php:415]: fixed.

File: src/Protector.php (hunk 2 above)
Comment (lupinitylabs) [reply src/Protector.php]: fixed

File: src/ProtectorServiceProvider.php [hunk 3]
```diff
@@ -41,13 +42,18 @@ public function boot(): void
...
         $this->mergeConfigFrom(__DIR__ . '/../config/protector.php', 'protector');
 
-        // Register the main class to use with the facade
+        // Register the main class to use with the facade.
         $this->app->singleton('protector', function () {
             return new Protector;
         });
+
+        // Register the SchemaState proxy classes.
+        $this->app->bind(MySqlSchemaStateProxy::class, function ($app, array $params) {
```
Comment (gael-connan-cybex) [inline src/ProtectorServiceProvider.php:54]: if this approach is supposed to be flexible or modular, wouldnt it make sense to have this configurable and loaded dynamically?

Comment (lupinitylabs) [reply src/ProtectorServiceProvider.php:54]: Yes, that's the idea. This is a POC that only implements MySql.
The goal would be to be able to bind a custom SchemaStateProxy to any SchemaState, including SchemaStates that might stem from third-party packages adding support for databases Laravel doesn't support.

Comment (gael-connan-cybex) [reply src/ProtectorServiceProvider.php:54]: fair nuff

File: src/Classes/MySqlSchemaStateProxy.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        $parameters = [
+            '--add-locks',
+            '--routines',
+            '--tz-utc',
+            '--column-statistics=0',
+            '--result-file="${:LARAVEL_LOAD_PATH}"',
+            '--max-allowed-packet='.$this->protector->getMaxPacketLength(),
+            ...array_keys(array_filter($conditionalParameters)),
+            '"${:LARAVEL_LOAD_DATABASE}"',
```
Comment (gael-connan-cybex) [inline src/Classes/MySqlSchemaStateProxy.php:66]: who interprets this wonderful syntax

**Review by gael-connan-cybex — APPROVED** [review]
tested

**Pull request merged**